packages/kit/src/runtime/client/utils.js

```diff
@@ -80,12 +80,59 @@ const levels = {
 
 /**
  * @param {Element} element
+ * @returns {Element | null}
+ */
+function parent_element(element) {
+	let parent = element.assignedSlot ?? element.parentNode;
+
+	// @ts-expect-error handle shadow roots
+	if (parent?.nodeType === 11) parent = parent.host;
+
+	return /** @type {Element} */ (parent);
+}
+
+/**
+ * @param {Element} element
+ * @param {Element} target
+ */
+export function find_anchor(element, target) {
+	while (element !== target) {
+		if (element.nodeName.toUpperCase() === 'A') {
+			return /** @type {HTMLAnchorElement | SVGAElement} */ (element);
+		}
+
+		element = /** @type {Element} */ (parent_element(element));
+	}
+}
```

Comment:
I've no doubt that's true, but unfortunately `closest()` doesn't work, because it can't find elements in parent shadow DOM — yet another reason adding it to the web platform was a mistake

Reply:
(incidentally, `element.closest(selector)` will return `element` if it matches `selector`, so if this _did_ work — if we decided not to bother with shadow DOM, on the basis that the wounds are self-inflicted — we could dispense with the first `if` block)

Reply:
> I've no doubt that's true, but unfortunately `closest()` doesn't work, because it can't find elements in parent shadow DOM — yet another reason adding it to the web platform was a mistake

Oh, makes sense. I didn't thought about that.


> (incidentally, `element.closest(selector)` will return `element` if it matches `selector`, so if this _did_ work — if we decided not to bother with shadow DOM, on the basis that the wounds are self-inflicted — we could dispense with the first `if` block)

I included this `element.nodeName.toUpperCase() !== 'A'` check because it was so much faster than `.clostest('a')` on an `HTMLAnchorElement` in my tests.